### PR TITLE
Grab bag of issues

### LINKED ISSUE #3373: Update install instructions to be more aligned with conda-forge
In light of #3369 and comments @ehsteve made in chat last week, we probably need to refresh our install instructions.

### LINKED ISSUE #3789: Document t in TimeRange
`TimeRange` had a `.__contains__` method, but this isn't well documented. We should add some more obvious documentation that one can do `t in TimeRange`.

### LINKED ISSUE #3820: Explain intersphinx linking in our developer guide
Unless I'm missing something, we don't really explain how to make use of intersphinx linking in our developer guide.  The closest I can find is: https://docs.sunpy.org/en/latest/dev_guide/documentation.html.

We should explain:

- Using backticks to link to SunPy stuff as well as external stuff
- Using the tilde to hide all but the last segment
- Using prefix qualifiers such as `:func:` and `:meth:` for functions and methods, respectively, so that they are rendered with the additional parentheses.

### LINKED ISSUE #4008: Missing metadata warnings aren't very helpful
I'm currently working on some code the raises a lot of Missing metadata warnings, that show up like this
```
WARNING: SunpyUserWarning: Missing metadata for observer: assuming Earth-based observer.
 [sunpy.map.mapbase]
```
This isn't very helpful though, as I don't know what line in my code, or object, is the cause of these warnings.

It would be good to add some extra information so the user knows what is causing the warnings, and therefore making them easier to diagnose and fix.

### LINKED ISSUE #4130: Update CONTRIBUTING.rst with a short primer on contributing to OSS
cc @Cadair and @dpshelio 

### Description
We may want to consider including the following in our contributor guidelines:

* A statement like: SunPy is open source, under a BSD-2 license, and by contributing you are stating that you have the right to and agree to have your work distributed under the terms of that license.
* Something about institutional copyright and OSS policy
* A link to resources about how to figure out if your institution has an OSS policy and allows you to contribute to OSS
* Who to ask if you don't know where to start

### LINKED ISSUE #4167: Eventually remove our minigallery Sphinx directive
The custom Sphinx directive to insert mini-galleries (#4124) has now been upstreamed to `sphinx-gallery` (sphinx-gallery/sphinx-gallery#685).  The version there is a bit more robust and featured.  Once it shows up in a release of `sphinx-gallery`, we can remove the version that we have in our code.

### LINKED ISSUE #4255: SunpyDeprecationWarnings should cause the doc build to fail
This is especially true in the gallery. I tried to set a filter in the `conf.py` file, but that didn't seem to propagate into the gallery.

### PR DESCRIPTION
**Pull request by nabobalis**
Fixes #3373
Fixes #4008
Fixes #3789
Fixes #4167
Fixes #4255
Fixes #4130
Fixes #3820 (with a patch to the theme repo as well https://github.com/sunpy/sunpy-sphinx-theme/pull/114)

TODO:
- [ ] Consider adding -test to our tox test jobs or coming up with better names
- [x] A page on how to read the output test logs: what to focus on, what to skip etc.

### CODE REVIEW

**Comment by dstansby** [comment]
What's the context/motivation for this?

**Comment by nabobalis** [comment]
The context is trying to see what happens if we unindent the retry code in the sample data to get RTD to  successfully do retry on the failed sample data. 

I need to rewrite more code here.

**Comment by nabobalis** [comment]
Seems we are failing to download the sunspot data, I am not sure why.

**Comment by nabobalis** [comment]
Some of this should be backport but will need to be done manually.

**Comment by nabobalis** [comment]
I plan to update the CI guide as well.

**Comment by nabobalis** [comment]
TODO:
-[ ] Add tox command for using the precommit

**Comment by nabobalis** [comment]
On the warning for examples I can get this:

```
Unexpected failing examples:
/home/nabil/GitHub/sunpy/examples/map/map_data_histogram.py failed leaving traceback:
Traceback (most recent call last):
  File "/home/nabil/GitHub/sunpy/.tox/build_docs/lib/python3.8/site-packages/sphinx_gallery/gen_gallery.py", line 159, in call_memory
    return 0., func()
  File "/home/nabil/GitHub/sunpy/.tox/build_docs/lib/python3.8/site-packages/sphinx_gallery/gen_rst.py", line 466, in __call__
    exec(self.code, self.fake_main.__dict__)
  File "/home/nabil/GitHub/sunpy/examples/map/map_data_histogram.py", line 24, in <module>
    aia_smap = aia.submap(bottom_left, top_right)
  File "/home/nabil/GitHub/sunpy/.tox/build_docs/lib/python3.8/site-packages/sunpy/util/decorators.py", line 320, in inner_f
    warnings.warn(f"Pass {', '.join(args_msg)} as keyword args. "
sunpy.util.exceptions.SunpyDeprecationWarning: Pass top_right=<SkyCoord (Helioprojective: obstime=2011-06-07T06:33:02.770, rsun=696000000.0 m, observer=<HeliographicStonyhurst Coordinate (obstime=2011-06-07T06:33:02.770): (lon, lat, radius) in (deg, deg, m)
    (-0.00406308, 0.04787238, 1.51846026e+11)>): (Tx, Ty) in arcsec
    (100., 400.)> as keyword args. From version 2.1 passing these as positional arguments will result in an error.
```

File: examples/computer_vision_techniques/loop_edge_enhance.py [hunk 1]
```diff
@@ -21,7 +21,7 @@
 aia = sunpy.map.Map(AIA_171_IMAGE)
 bottom_left = SkyCoord(750 * u.arcsec, -200 * u.arcsec, frame=aia.coordinate_frame)
 top_right = SkyCoord(1500 * u.arcsec, 550 * u.arcsec, frame=aia.coordinate_frame)
-aia_smap = aia.submap(bottom_left, top_right=top_right)
+aia_smap = aia.submap(bottom_left, top_right)
```
Comment (nabobalis) [inline examples/computer_vision_techniques/loop_edge_enhance.py:24]: I've done this to make sure the warning is treated as an error.
This should be undone.

**Comment by nabobalis** [comment]
Now I can't get it to work again.

**Comment by nabobalis** [comment]
Still getting:
```python
DEBUG: Failed to download EVE_TIMESERIES from http://data.sunpy.org/sunpy/v1/20110607_EVE_L0CS_DIODES_1m.txt: Timeout on reading data from socket [sunpy.data._sample]
loading translations [en]... done

Task was destroyed but it is pending!
task: <Task pending name='Task-120' coro=<Downloader._write_worker() running at /home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4331/lib/python3.8/site-packages/parfive/downloader.py:469> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7f44e4d46250>()]>>
Exception ignored in: <coroutine object Downloader._write_worker at 0x7f44e4d441c0>
Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4331/lib/python3.8/site-packages/parfive/downloader.py", line 469, in _write_worker
    offset, chunk = await queue.get()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4331/lib/python3.8/asyncio/queues.py", line 165, in get
    getter.cancel()  # Just in case getter is not done yet.
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4331/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/home/docs/checkouts/readthedocs.org/user_builds/sunpy/conda/4331/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
making output directory... done
```

File: docs/dev_guide/pr_review_procedure.rst [hunk 2]
```diff
@@ -37,15 +37,35 @@ Continuous Integration
...
+* `pep8speaks <https://github.com/OrkoHunter/pep8speaks>`_: Performs a PEP8 check on any submitted code. This is updated as the code changes.
 
-* Giles: Returns a link if the documentation builds successfully.
+Checks (they are located at the bottom of the pull request):
 
-* `Azure Pipelines <https://dev.azure.com/sunpy/sunpy/_build>`_: Runs our test suite on all three operating systems.
+* `figure-tests (CircleCi) <https://circleci.com/gh/sunpy/sunpy/>`_: Runs two figure tests environments ("py38-figure", "py37-figure-devdeps").
 
-* `CodeCov <https://codecov.io/gh/sunpy/sunpy/>`_: Checks how many lines of the code lack test coverage.
+* figure_report/figure_tests (Giles): Show the final results of the figure tests.
```
Comment (nabobalis) [inline docs/dev_guide/pr_review_procedure.rst:48]: I just noticed when writing this, we have three figure giles checks.

File: docs/dev_guide/documentation.rst [hunk 3]
```diff
@@ -10,25 +10,79 @@ Overview
...
+
+If you link to a method of a class, it will work.
+
+.. code-block:: rst
+
+    :meth:`numpy.ndarray.mean`
+
+:meth:`numpy.ndarray.mean` but ``:func:`` will not.
+
+Never add ``:class:`` this will break the linking.
```
Comment (nabobalis) [inline docs/dev_guide/documentation.rst:85]: Much thanks to the MPL docs for this section.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -72,10 +72,10 @@ repos:
         args: ['--in-place', '--remove-all-unused-imports', '--remove-unused-variable']
         exclude: ".*(.fits|.fts|.fit|.txt|tca.*|extern.*|.rst|.md|__init__.py|sunpy/extern|docs/conf.py)$"
   - repo: https://github.com/timothycrosley/isort
-    rev: 4.3.21-2
+    rev: 5.1.0
```
Comment (nabobalis) [inline .pre-commit-config.yaml:75]: isort update.

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -84,6 +84,16 @@ The main one is the SunPy repository with where all the known `issues`_ with Sun
...
 
+.. note::
+
+    sunpy is open source, under the BSD-2 license, and by contributing you are stating that you have the right to and agree to, have your work distributed under the terms of this license.
+
+    If you work at a university or research institution, you will need to check if you can contribute code you have written at work.
+    You should start by checking if there is a Open Source Software Policy (e.g., `Standford's policy <https://otl.stanford.edu/open-source-stanford>`__) for your work place.
+    If not, `OSS-Watch <http://oss-watch.ac.uk/resources/contributing>`__ summaries what you will need to check and who to ask, however this is from a UK point of view.
+    `Standford's guidance <https://otl.stanford.edu/sites/g/files/sbiybj10286/f/otlcopyrightguide.pdf>`__ for example allows someone to contribute and open source their code.
+    If you are unsure if your university or institution allows you to contribute under the BSD-2 license, you should contact the relevant department or administrator that deals with copyright.
```
Comment (nabobalis) [inline CONTRIBUTING.rst:95]: This needs more passes and reviews. I was unsure what to suggest. I was able to find some resources but not many.

**Comment by nabobalis** [comment]
I might need to split this up.

**Comment by nabobalis** [comment]
Dev build
```
INFO: Failed to download one or more sample data files, retrying with a mirror. [sunpy.data._sample]
Traceback (most recent call last):
  File "/home/vsts/work/1/s/examples/acquiring_data/2011_06_07_sampledata_overview.py", line 32, in <module>
    aia_cutout03_map = sunpy.map.Map(sample_data.AIA_193_CUTOUT03_IMAGE)
AttributeError: module 'sunpy.data.sample' has no attribute 'AIA_193_CUTOUT03_IMAGE'

Task was destroyed but it is pending!
task: <Task pending name='Task-131' coro=<Downloader._write_worker() running at /home/vsts/work/1/s/.tox/py38-devdeps/lib/python3.8/site-packages/parfive/downloader.py:469> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x7fd3e64d3f70>()]>>
Exception ignored in: <coroutine object Downloader._write_worker at 0x7fd3e773e6c0>
Traceback (most recent call last):
  File "/home/vsts/work/1/s/.tox/py38-devdeps/lib/python3.8/site-packages/parfive/downloader.py", line 469, in _write_worker
    offset, chunk = await queue.get()
  File "/opt/hostedtoolcache/Python/3.8.3/x64/lib/python3.8/asyncio/queues.py", line 165, in get
    getter.cancel()  # Just in case getter is not done yet.
  File "/opt/hostedtoolcache/Python/3.8.3/x64/lib/python3.8/asyncio/base_events.py", line 719, in call_soon
    self._check_closed()
  File "/opt/hostedtoolcache/Python/3.8.3/x64/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
ERROR: InvocationError for command /bin/bash -ec 'for f in /home/vsts/work/1/s/examples/*/*.py; do [[ $f == *skip* ]] && continue; echo $f && python $f; done' (exited with code 1)
py38-devdeps: Shutting down tox-pypi-filter server
```

**Pull request closed without merging**